python/search_completer/hierarchical_search_completer.py

```diff
@@ -0,0 +1,187 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import sgtk
+from sgtk.platform.qt import QtCore, QtGui
+
+from .search_completer import SearchCompleter
+
+shotgun_model = sgtk.platform.import_framework("tk-framework-shotgunutils", "shotgun_model")
+
+
+class HierarchicalSearchCompleter(SearchCompleter):
+    """
+    A standalone :class:`PySide.QtGui.QCompleter` class for matching SG entities to typed text.
+
+    :signal: ``node_activated(str, int, str, str, list)`` - Fires when someone activates a
+        node inside the search results. The parameters are ``type``, ``id``, ``name``,
+        ``label path`` and ``incremental_paths``. If the node activated is not an entity,
+        ``type`` and ``id`` will be ``None``.
+
+    :modes: ``MODE_LOADING, MODE_NOT_FOUND, MODE_RESULT`` - Used to identify the
+        mode of an item in the completion list
+
+    :model role: ``MODE_ROLE`` - Stores the mode of an item in the completion
+        list (see modes above)
+
+    :model role: ``SG_DATA_ROLE`` - Role for storing shotgun data in the model
+    """
+
+    # path label, entity type, entity id, name, incremental path
+    node_activated = QtCore.Signal(str, int, str, str, list)
+
+    def __init__(self, parent=None):
+        """
+        :param parent: Parent widget
+        :type parent: :class:`~PySide.QtGui.QWidget`
+        """
+        super(HierarchicalSearchCompleter, self).__init__(parent)
+        self.set_search_root(self._bundle.context.project)
+
+    def set_search_root(self, entity):
+        """
+        Allows to change the root of the search.
+
+        :param dict entity: Entity to search under. If ``None``, the search will be done
+            at the site level. Note that only ``Project`` entities are supported at the moment.
+        """
+        if not entity:
+            self._search_root = "/"
+        else:
+            self._search_root = "/Project/%d" % entity.get("id")
```

Comment:
@daigles has suggested that in order to keep the navigation API as flexible as possible, we should not do any string manipulation of these paths at all - think of them as black box tokens being sent down from the server (i think of them as guids). Instead, pass the entity into the new resolve method to get paths back and then we can use the first path that is returned. That way, the client side is completely agnostic and all the resolve logic stays on the server side! in case there is more than one value returned we should log it (i don't think it's a warning but useful for debugging at the very least)